distributed/scheduler.py

```diff
@@ -654,6 +657,79 @@ def _to_dict_no_nest(self, *, exclude: Container[str] = ()) -> dict[str, Any]:
             members=True,
         )
 
+    def add_to_processing(self, ts: TaskState) -> None:
+        """Assign a task to this worker for compute."""
+        assert self.scheduler_ref and (scheduler := self.scheduler_ref())
+        tg = ts.group
+        self.task_groups_count[tg.name] += 1
+        scheduler.task_groups_count_global[tg.name] += 1
+        self.processing.add(ts)
+        for dts in ts.dependencies:
+            if self not in dts.who_has:
+                self._inc_needs_replica(dts)
+
+    def remove_from_processing(self, ts: TaskState) -> None:
+        """Remove a task from a workers processing"""
+        if ts not in self.processing:
+            return
+        assert self.scheduler_ref and (scheduler := self.scheduler_ref())
+        self.task_groups_count[ts.group.name] -= 1
+        scheduler.task_groups_count_global[ts.group.name] -= 1
+        if not self.task_groups_count[ts.group.name]:
+            del self.task_groups_count[ts.group.name]
+        if not scheduler.task_groups_count_global[ts.group.name]:
+            del scheduler.task_groups_count_global[ts.group.name]
+        self.processing.remove(ts)
+        for dts in ts.dependencies:
+            if dts in self.needs_what:
+                self._dec_needs_replica(dts)
+
+    def remove_replica(self, ts):
+        """The worker no longer has a task in memory"""
+        if ts in self.needs_what:
+            self._inc_needs_replica(ts)
+        self.nbytes -= ts.get_nbytes()
+        del self._has_what[ts]
+        ts.who_has.remove(self)
+
+    def _inc_needs_replica(self, ts: TaskState) -> None:
+        """Assign a task to this worker and update occupancies"""
+        assert self.scheduler_ref and (scheduler := self.scheduler_ref())
+        if ts not in self.needs_what:
+            self.needs_what[ts] = 1
+            nbytes = ts.get_nbytes()
+            self._network_occ += nbytes
+            scheduler._network_occ_global += nbytes
+        else:
+            self.needs_what[ts] += 1
+
+    def _dec_needs_replica(self, ts: TaskState) -> None:
+        assert self.scheduler_ref and (scheduler := self.scheduler_ref())
+        if ts in self.needs_what:
+            self.needs_what[ts] -= 1
+            if not self.needs_what[ts]:
+                del self.needs_what[ts]
+                nbytes = ts.get_nbytes()
+                self._network_occ -= nbytes
+                scheduler._network_occ_global -= nbytes
+
+    def add_replica(self, ts: TaskState) -> None:
+        """The worker acquired a replica of task"""
+        assert self.scheduler_ref and (scheduler := self.scheduler_ref())
+        nbytes = ts.get_nbytes()
+        if ts in self.needs_what:
+            del self.needs_what[ts]
```

Comment:
I think there might be an issue with the removal of `self.needs_what[ts]` here and only incrementing it by one on `remove_replica`.

Reply:
`needs_what` is a counter for how many tasks assigned to this worker require this particular key. As soon as we call `add_replica` this counter drops immediately to zero.

`needs_what` keys are disjoint with `has_what`